README.md
# hex-text

`hex-text` is a small library for converting between `ByteString`s and their representations as hexidecimal numbers encoded as `Text`.

# Motivation

When using Stripe for payments, Stripe sends a signature as a hexidecimal `Text` value. The `cryptonite` package can be used to verify the signature, but it requires `ByteString` values, not `Text`. 

# Example usage

A `ByteString` is a list of bytes. A byte is a number between 0 and 255, represented by the `Word8` type. In a fixed-width hexidecimal representation, the lowest byte 0 is represented by the hex string `00`, and the greatest byte 255 is represented by the hex string `ff`. So, for example, the `ByteString` consisting of bytes \[ 1, 2, 3, 253, 254, 255 \] is represented as `010203fdfeff`.

```haskell
λ> import Text.Hex (encodeHex)
λ> import Data.ByteString (pack)

λ> (encodeHex . pack) [1, 2, 3, 253, 254, 255]
"010203fdfeff"
```



hex-text/changelog.md
### 0.1.0.8 (2022-12-31)

Switch testing to `hspec`

Raise minimum required GHC to 8.10

### 0.1.0.7 (2022-11-10)

Start of changelog


hex-text/test/Main.hs
import Text.Hex (decodeHex, encodeHex, lazilyEncodeHex)

import Test.Hspec (hspec, describe, it, shouldBe, Spec)

import Prelude ((.), ($), cycle, IO, fmap, Maybe (..))
import Data.String (fromString)

import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Lazy as LazyByteString
import qualified Data.Text as Text
import qualified Data.Text.Lazy as LazyText

main :: IO ()
main = hspec $ do
    encodeHexTests
    decodeHexTests
    lazilyEncodeHexTests

encodeHexTests :: Spec
encodeHexTests = describe "encodeHex" $ do
    it "can encode a single byte" $ do
        let f = encodeHex . ByteString.singleton
        f 192 `shouldBe` fromString "c0"
        f 168 `shouldBe` fromString "a8"
    it "can encode many bytes" $ do
        let f = encodeHex . ByteString.pack
        f [192, 168, 1, 2] `shouldBe` fromString "c0a80102"

decodeHexTests :: Spec
decodeHexTests = describe "decodeHex" $ do
    let f = fmap ByteString.unpack . decodeHex . Text.pack
    it "can decode" $ do
        f "c0a80102" `shouldBe` Just [192,168,1,2]
        f "C0A80102" `shouldBe` Just [192,168,1,2]
    it "can fail" $ do
        f "c0a8010" `shouldBe` Nothing
        f "x0a80102" `shouldBe` Nothing

lazilyEncodeHexTests :: Spec
lazilyEncodeHexTests = describe "lazilyEncodeHex" $ do
    let f = lazilyEncodeHex . LazyByteString.pack
    it "can decode part of an infinite list" $
        (LazyText.take 8 . f . cycle) [1, 2, 3] `shouldBe` fromString "01020301"


hex-text/test/test.hs
import Data.Foldable
import Data.Semigroup
import Data.String
import Numeric.Natural
import System.Exit
import Text.Hex

import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Lazy as LazyByteString
import qualified Data.Text as Text
import qualified Data.Text.Lazy as LazyText

main = run tests

tests = encodeHexTests <> decodeHexTests <> lazilyEncodeHexTests

run [] = putStrLn "Okay"
run xs = die $ "Test failures: " <> show (xs :: [Natural])

infix 0 #
x # True = []
x # False = [x]

encodeHexTests = fold
  [ 1 # (encodeHex . ByteString.singleton) 192
      == fromString "c0"
  , 2 # (encodeHex . ByteString.singleton) 168
      == fromString "a8"
  , 3 # (encodeHex . ByteString.pack) [192, 168, 1, 2]
      == fromString "c0a80102"
  ]

decodeHexTests = fold
  [ 4 # (fmap ByteString.unpack . decodeHex . Text.pack) "c0a80102"
      == Just [192,168,1,2]
  , 5 # (fmap ByteString.unpack . decodeHex . Text.pack) "c0a8010"
      == Nothing
  , 6 # (fmap ByteString.unpack . decodeHex . Text.pack) "x0a80102"
      == Nothing
  , 7 # (fmap ByteString.unpack . decodeHex . Text.pack) "C0A80102"
      == Just [192,168,1,2]
  ]

lazilyEncodeHexTests = fold
  [ 8 # (LazyText.take 8 . lazilyEncodeHex . LazyByteString.pack . cycle) [1, 2, 3]
      == fromString "01020301"
  ]


hex-text/test/doctest.hs
import Test.DocTest

main :: IO ()
main =
  doctest [ "-isrc", "src/Text/Hex.hs" ]


hex-text/src/Text/Hex.hs
module Text.Hex
    (
    -- * Encoding and decoding
      encodeHex
    , decodeHex
    , lazilyEncodeHex

    -- * Types
    , Text
    , LazyText
    , ByteString
    , LazyByteString

    -- * Type conversions
    , lazyText
    , strictText
    , lazyByteString
    , strictByteString

    ) where

import Prelude (either, Maybe (..), const)

-- base16-bytestring
import qualified Data.ByteString.Base16 as Base16
import qualified Data.ByteString.Base16.Lazy as LazyBase16

-- bytestring
import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Lazy as LazyByteString

-- text
import qualified Data.Text as Text
import qualified Data.Text.Encoding as Text
import qualified Data.Text.Lazy as LazyText
import qualified Data.Text.Lazy.Encoding as LazyText

-- | Strict byte string

type ByteString =
    ByteString.ByteString

-- | Lazy byte string

type LazyByteString =
    LazyByteString.ByteString

-- | Strict text

type Text =
    Text.Text

-- | Lazy text

type LazyText =
    LazyText.Text

-- |
-- Encodes a byte string as hexidecimal number represented in text.
-- Each byte of the input is converted into two characters in the
-- resulting text.
--
-- >>> (encodeHex . ByteString.singleton) 192
-- "c0"
--
-- >>> (encodeHex . ByteString.singleton) 168
-- "a8"
--
-- >>> (encodeHex . ByteString.pack) [192, 168, 1, 2]
-- "c0a80102"
--
-- 'Text' produced by @encodeHex@ can be converted back to a
-- 'ByteString' using 'decodeHex'.
--
-- The lazy variant of @encodeHex@ is 'lazilyEncodeHex'.

encodeHex :: ByteString -> Text
encodeHex bs =
    Text.decodeUtf8 (Base16.encode bs)

-- |
-- Decodes hexidecimal text as a byte string. If the text contains
-- an even number of characters and consists only of the digits @0@
-- through @9@ and letters @a@ through @f@, then the result is a
-- 'Just' value.
--
-- Unpacking the ByteString in the following examples allows for
-- prettier printing in the REPL.
--
-- >>> (fmap ByteString.unpack . decodeHex . Text.pack) "c0a80102"
-- Just [192,168,1,2]
--
-- If the text contains an odd number of characters, decoding fails
-- and produces 'Nothing'.
--
-- >>> (fmap ByteString.unpack . decodeHex . Text.pack) "c0a8010"
-- Nothing
--
-- If the text contains non-hexidecimal characters, decoding fails
-- and produces 'Nothing'.
--
-- >>> (fmap ByteString.unpack . decodeHex . Text.pack) "x0a80102"
-- Nothing
--
-- The letters may be in either upper or lower case. This next
-- example therefore gives the same result as the first one above:
--
-- >>> (fmap ByteString.unpack . decodeHex . Text.pack) "C0A80102"
-- Just [192,168,1,2]

decodeHex :: Text -> Maybe ByteString
decodeHex txt =
    either (const Nothing) Just (Base16.decode (Text.encodeUtf8 txt))

-- |
-- @lazilyEncodeHex@ is the lazy variant of 'encodeHex'.
--
-- With laziness, it is possible to encode byte strings of
-- infinite length:
--
-- >>> (LazyText.take 8 . lazilyEncodeHex . LazyByteString.pack . cycle) [1, 2, 3]
-- "01020301"

lazilyEncodeHex :: LazyByteString -> LazyText
lazilyEncodeHex bs =
    LazyText.decodeUtf8 (LazyBase16.encode bs)

lazyText :: Text -> LazyText
lazyText = LazyText.fromStrict

strictText :: LazyText -> Text
strictText = LazyText.toStrict

lazyByteString :: ByteString -> LazyByteString
lazyByteString = LazyByteString.fromStrict

strictByteString :: LazyByteString -> ByteString
strictByteString = LazyByteString.toStrict
